blog/content/post/2024-03-15-uefi-sb-implementation-mpl.md

```diff
@@ -0,0 +1,361 @@
+methods have been omitted due to their logic, assuming that firmware can remove
+certificates enrolled this way from the UEFI BIOS Menu.
+
+The conclusions allowed us to proceed with UEFI Secure Boot integration in the
+Yocto layer.
+
+## Integrating Secure Boot into a Yocto layer
+
+Integrating UEFI Secure Boot into an existing Yocto layer is possible by using
+the [meta-secure-core](https://github.com/Wind-River/meta-secure-core) layer in
+your build. Its sublayer – `meta-efi-secure-boot` introduces mechanisms that
+allow verifying various files used in the boot process. It offers two secure
+boot technologies:
+
+- UEFI Secure Boot, which verifies images loaded by UEFI firmware against
+certificates
+- MOK (Machine Owner Key) Secure Boot, which extends UEFI SB by
+introducing user-added Machine Owner Keys
+
+### Custom certificates
+
+UEFI Secure Boot allows the user to enroll custom certificates which will be
+used to verify images. The certificates can be loaded automatically thanks to
+the
+[Automatic Certificate Provisioning](https://github.com/Wind-River/meta-secure-core/tree/master/meta-efi-secure-boot#automatic-certificate-provision)
+procedure provided by `meta-secure-core`. The procedure uses `LockDown.efi`,
+which can only be executed when UEFI Secure Boot is disabled. `LockDown.efi`
+provisions the new certificates, which are built into it during build.
+
+UEFI Secure Boot uses the DB certificate to verify GRUB. Aside from that,
+`grubenv` and `grub.cfg` are signed during build using a boot key, which can
+be defined in the layer. GRUB then tries to load the kernel, which is also
+checked against DB. If any of the checks fail, the respective file cannot be
+booted. This mechanism increases the system's security by making sure that the
+firmware does not boot untrusted files.
+
+### Implementation
+
+To integrate the mechanisms available in `meta-secure-core` you need to
+integrate that layer into your build. We use
+[kas-container](https://github.com/siemens/kas/blob/master/kas-container) to set
+up bitbake projects.
+
+In your kas configuration file add the following to the `repos` section:
+
+```yaml
+meta-secure-core:
+url: https://github.com/Wind-River/meta-secure-core.git
+refspec: 8dc9f1b4a735eccee65a8896760e473e110d147e
+layers:
+    meta-efi-secure-boot:
+    meta:
+    meta-signing-key:
+```
+
+Define `meta-secure-core` variables in your layer's `local.conf`:
+
+```conf
+# UEFI Secure Boot variables
+# Use only UEFI Secure Boot without MOK Secure Boot
+UEFI_SELOADER = "0"
+GRUB_SIGN_VERIFY = "1"
+UEFI_SB = "1"
+MOK_SB = "0"
+
+# We want grub-efi from meta-efi-secure-boot to install bootfiles under
+# /boot/EFI/BOOT
+EFI_BOOT_PATH = "/boot/EFI/BOOT"
+
+DISTRO_FEATURES_NATIVE:append = " efi-secure-boot"
+DISTRO_FEATURES:append = " efi-secure-boot modsign"
+MACHINE_FEATURES_NATIVE:append = " efi"
+MACHINE_FEATURES:append = " efi"
+
+DEBUG_FLAGS:forcevariable = ""
+IMAGE_INSTALL:append = " kernel-image-bzimage"
+```
+
```

Comment:
Is this okay? ad87522d2cb6e682dc5e76ebb3c26599f8ec764c

Reply:
@PLangowski what flags you had in mind? Maybe we could explain variables like GRUB_SIGN_VERIFY, UEFI_SB etc? Also, it would be easier if you would point to a specifc line, like https://github.com/3mdeb/news-and-ideas/commit/ad87522d2cb6e682dc5e76ebb3c26599f8ec764c#diff-408ae19700697ee529e7794c3865363586a3ffe06152e4c90cb578ba42221b40R227

Reply:
https://github.com/3mdeb/news-and-ideas/pull/590/commits/6dc32699ca057014941a019ba6d503e89e783c3e#diff-408ae19700697ee529e7794c3865363586a3ffe06152e4c90cb578ba42221b40R279